File: src/wfrp-helper-web/src/model/crafting/Resource.js
import Entity, {CreateColumn, CreateFormField, RegisterEntity} from "../Entity";
import {FormFieldType} from "../../form/FormFieldType";
import {availabilityEntityName} from "./Availability";

export default class Resource extends Entity {
    static entityName: string = 'resource';

    name: string;
    description: string;
    priceMultiplier: number;
    weight: number;
    durability: number;
    strength: number;
    availability: string = 'COMMON';
}

RegisterEntity(Resource, [
    CreateColumn('name', FormFieldType.TEXT),
    CreateColumn('description', FormFieldType.TEXT_AREA),
    CreateColumn('priceMultiplier', FormFieldType.FLOAT),
    CreateColumn('durability', FormFieldType.FLOAT),
    CreateColumn('strength', FormFieldType.FLOAT),
    CreateColumn('availability', FormFieldType.TEXT),
], [
    CreateFormField('name', FormFieldType.TEXT),
    CreateFormField('description', FormFieldType.TEXT_AREA),
    CreateFormField('priceMultiplier', FormFieldType.FLOAT),
    CreateFormField('weight', FormFieldType.FLOAT),
    CreateFormField('durability', FormFieldType.FLOAT),
    CreateFormField('strength', FormFieldType.FLOAT),
    CreateFormField('availability', FormFieldType.ENUM_SELECT, availabilityEntityName),
]);

File: src/main/java/pl/khuzzuk/wfrp/helper/model/crafting/resource/Resource.java
package pl.khuzzuk.wfrp.helper.model.crafting.resource;

import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.NaturalId;
import org.hibernate.annotations.Type;
import org.hibernate.validator.constraints.Length;
import pl.javahello.RemoteEntity;
import pl.javahello.RemoteEntity.SecuredService;
import pl.khuzzuk.wfrp.helper.common.EnumType;
import pl.khuzzuk.wfrp.helper.repo.BaseEntity;
import pl.khuzzuk.wfrp.helper.model.crafting.inventory.Availability;

import javax.persistence.*;
import javax.validation.constraints.NotNull;

@Getter
@Setter
@EqualsAndHashCode(of = "name", callSuper = false)
@Entity
@Table(schema = "crafting")
@RemoteEntity(transactional = true, stomp = true)
@SecuredService(allowRead = true)
public class Resource extends BaseEntity {
    @Id
    @SequenceGenerator(name = "resource_seq_gen", schema = "crafting", sequenceName = "resource_id_seq", allocationSize = 1)
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "resource_seq_gen")
    private Long id;
    @NaturalId
    private @Length(min = 3, max = 100) String name;
    private @Length(max = 500) String description;
    private float priceMultiplier = 1;
    private float weight;
    private float durability;
    private float strength;
    @Type(type = EnumType.DEF)
    private @NotNull Availability availability = Availability.COMMON;
}
